### ISSUE TITLE
GFF output format causes parsing problems for other programs.

### ISSUE THREAD

**Issue by conchoecia** [oushujun/EDTA#83]
Hi @oushujun,

I noticed that in the GFF file all of the individual GFF entries that belong to the same TE family have the same `ID=` field. This makes IGV and other parsing software ignore the majority of GFF entries because many lines have the same `ID=`. To get around this, I renamed all the `ID=` fields to `TEID=` and prepended the line number.

You can see that when I loaded up the original GFF (top), many of the annotations don't appear. The track in which I renamed everything (bottom), all of the entries are displayed.
<img width="1129" alt="Screen Shot 2020-05-12 at 10 30 03 AM" src="https://user-images.githubusercontent.com/3123273/81726213-e0940600-943b-11ea-89a3-7262efa0a2d2.png">

Here's what the original GFF looks like:

```
##gff-version 2
##date Tue May 12 06:19:14 UTC 2020
##Chromosome Annotator Repeat_class/superfamily Start End Score Strand Phase Annotation
c1      EDTA    LTR/unknown     66      200     .       -       .       ID=TE_00000472_INT;Identity=0.911;Method=homology
c1      EDTA    LTR/unknown     250     504     .       -       .       ID=TE_00000472_INT;Identity=0.903;Method=homology
c1      EDTA    LTR/unknown     453     578     .       +       .       ID=TE_00000411_LTR;Identity=0.939;Method=homology
c1      EDTA    LTR/unknown     549     753     .       -       .       ID=TE_00000472_INT;Identity=0.907;Method=homology
```

Here's what the edited GFF looks like:

```
##gff-version 2
##date Tue May 12 06:19:14 UTC 2020
##Chromosome Annotator Repeat_class/superfamily Start End Score Strand Phase Annotation
c1      EDTA    LTR/unknown     66      200     .       -       .       ID=TE_annot_0;TEID=TE_00000472_INT;Identity=0.911;Method=homology
c1      EDTA    LTR/unknown     250     504     .       -       .       ID=TE_annot_1;TEID=TE_00000472_INT;Identity=0.903;Method=homology
c1      EDTA    LTR/unknown     453     578     .       +       .       ID=TE_annot_2;TEID=TE_00000411_LTR;Identity=0.939;Method=homology
c1      EDTA    LTR/unknown     549     753     .       -       .       ID=TE_annot_3;TEID=TE_00000472_INT;Identity=0.907;Method=homology
```

Would you consider using a naming scheme like I've proposed above to allow EDTA's GFF to easily be folded into downstream applications?

**Comment by oushujun** [comment]
HI @conchoecia,

Sorry for the delayed response. Thank you for your suggestion. Could you PR your script to this repo? I can add a conversion step to the final gff file.

Best,
Shujun

**Comment by Juke34** [comment]
I agree with @conchoecia the GFF output should be written in a better way. 
I see GFF2 in the header, you should use GFF3. 
You should use uniq ID by feature. If several features are linked together use ID/Parent relationship or use a common tag (locus_tag) between those linked features (as often done by JGI or NCBI).
instead of `LTR/unknown` in the 3rd column use a value accepted by the Sequence Ontology Consortium (as asked by the GFF3 specification). You can add as many attribute you wish in the 9th column e.g.  repeat_type="LTR/unknown"

**Comment by oushujun** [comment]
Hi @conchoecia and @Juke34,

Thank you very much for your suggestions. I am working to reformat the GFF3 output following the GFF3 standard [guideline](https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md). It involves reengineering a number of scripts in different programs so it will take some time. Below is a sample of the updated format for intact LTR-RTs. Please take a look if it follows the expected format and I can further adjust the rest of the scripts.

```
##gff-version 3
##date Wed May 20 22:34:48 UTC 2020
##seqid source repeat_class/superfamily start end score strand phase attributes
##sequence-region   Tzi8_chr1 1 307421186
Tzi8_chr1       LTR_retriever   repeat_region   54481   63973   .       -       .       ID=repeat_region1;Sequence_ontology=SO:0000657;Name=Tzi8_chr1:54486..63968#LTR/Copia;motif=TGCA;tsd=GAGTG;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   target_site_duplication 54481   54485   .       -       .       ID=lTSD_1;Parent=repeat_region1;Sequence_ontology=SO:0000434;Name=Tzi8_chr1:54486..63968#LTR/Copia;motif=TGCA;tsd=GAGTG;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   long_terminal_repeat    54486   55806   .       -       .       ID=lLTR_1;Parent=repeat_region1;Sequence_ontology=SO:0000286;Name=Tzi8_chr1:54486..63968#LTR/Copia;motif=TGCA;tsd=GAGTG;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   LTR/Copia       54486   63968   .       -       .       ID=LTRRT_1;Parent=repeat_region1;Sequence_ontology=SO:0000186;Name=Tzi8_chr1:54486..63968#LTR/Copia;motif=TGCA;tsd=GAGTG;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   long_terminal_repeat    62647   63968   .       -       .       ID=rLTR_1;Parent=repeat_region1;Sequence_ontology=SO:0000286;Name=Tzi8_chr1:54486..63968#LTR/Copia;motif=TGCA;tsd=GAGTG;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   target_site_duplication 63969   63973   .       -       .       ID=rTSD_1;Parent=repeat_region1;Sequence_ontology=SO:0000434;Name=Tzi8_chr1:54486..63968#LTR/Copia;motif=TGCA;tsd=GAGTG;ltr_identity=1.0000;Method=structural
###
Tzi8_chr1       LTR_retriever   repeat_region   146557  156096  .       -       .       ID=repeat_region2;Sequence_ontology=SO:0000657;Name=Tzi8_chr1:146565..156091#LTR/Copia;motif=TGCA;tsd=CTCCT;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   target_site_duplication 146557  146561  .       -       .       ID=lTSD_2;Parent=repeat_region2;Sequence_ontology=SO:0000434;Name=Tzi8_chr1:146565..156091#LTR/Copia;motif=TGCA;tsd=CTCCT;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   long_terminal_repeat    146565  147888  .       -       .       ID=lLTR_2;Parent=repeat_region2;Sequence_ontology=SO:0000286;Name=Tzi8_chr1:146565..156091#LTR/Copia;motif=TGCA;tsd=CTCCT;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   LTR/Copia       146565  156091  .       -       .       ID=LTRRT_2;Parent=repeat_region2;Sequence_ontology=SO:0000186;Name=Tzi8_chr1:146565..156091#LTR/Copia;motif=TGCA;tsd=CTCCT;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   long_terminal_repeat    154768  156091  .       -       .       ID=rLTR_2;Parent=repeat_region2;Sequence_ontology=SO:0000286;Name=Tzi8_chr1:146565..156091#LTR/Copia;motif=TGCA;tsd=CTCCT;ltr_identity=1.0000;Method=structural
Tzi8_chr1       LTR_retriever   target_site_duplication 156092  156096  .       -       .       ID=rTSD_2;Parent=repeat_region2;Sequence_ontology=SO:0000434;Name=Tzi8_chr1:146565..156091#LTR/Copia;motif=TGCA;tsd=CTCCT;ltr_identity=1.0000;Method=structural
###
```
Best,
Shujun

**Comment by Juke34** [comment]
Sounds good except that `LTR/Copia` is not part of the Sequence Ontology (SO). You say it is SO:0000186, which corresponds to LTR_retrotransposon.

**Comment by oushujun** [comment]
@Juke34 Yeah the SO system does not have TE superfamily classifications. So no SO for copia/gypsy/mutator/hAT/etc. So between obligately following the SO standard and retaining information, I choose the latter. Do you have a better way?

**Comment by Juke34** [comment]
Yes right it is not always easy to follow rigorously the specifications. You can do like that but 
go there http://www.sequenceontology.org and make a request for the missing term.
Otherwise you  can use a more generic term and specify then `LTR/Copia` in an attribute of the 9th column.

**Comment by oushujun** [comment]
@Juke34 Thanks for the suggestions. I have submitted a SO request and will see how it goes.

**Comment by conchoecia** [comment]
Thanks for pursuing this, @oushujun ! I think that using a generic term for field 2 and an attribute in field 9, like @Juke34 suggested, is a good fix.

**Comment by njaupan** [comment]
Hi, you can add sequential numbers based on ID at the 9th column,here is my script
 `cat EDTA.TEanno.gff |awk -v OFS='\t' '{gsub(";","\t")} {a[$9]++;$9=$9"_"a[$9]}1' |awk -v OFS='\t' '{print$1,$2,$3,$4,$5,$6,$7,$8,$9";"$10";"$11}' >EDTA.TEanno.fic.gff`
![Screenshot from 2020-06-22 09-50-50](https://user-images.githubusercontent.com/8072119/85262380-fb5f8e80-b46d-11ea-9af3-2331dfc3be2b.png)

**Comment by oushujun** [comment]
Hello all,

Thank you for waiting patiently. I just released a new version of EDTA (v1.9.0) which takes care of the gff3 output format. To update your results to this version, you don't need to rerun the entire pipeline. There is a patch script `EDTA/util/patch_1.8.3_to_1.9.0.pl` in the new package. Just execute that on your existing annotation. The patch should be done within an hour or so. Please let me know if you encounter any issues.

Best,
Shujun
